src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,164 @@
+/* Copyright (c) 2019 FIRST. All Rights Reserved.                             */
+/* Open Source Software - may be modified and shared by FRC teams. The code   */
+/* must be accompanied by the FIRST BSD license file in the root directory of */
+/* the project.                                                               */
+/*----------------------------------------------------------------------------*/
+
+package frc.robot.subsystems;
+
+import com.revrobotics.CANPIDController;
+import com.revrobotics.ControlType;
+import com.revrobotics.CANSparkMaxLowLevel.MotorType;
+import com.revrobotics.CANEncoder;
+import edu.wpi.first.wpilibj.DoubleSolenoid;
+import edu.wpi.first.wpilibj.DoubleSolenoid.Value;
+import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants;
+import frc.robot.Game;
+import frc.robot.MotorController;
+import frc.robot.Robot;
+
+public class ShooterSubsystem extends SubsystemBase {
+  /**
+   * Creates a new ShooterSubsystem.
+   */
+  private final MotorController mShooterMotor;
+  private final DoubleSolenoid mHoodDoubleSolenoid;
+  private final CANPIDController mShooterPID;
+  private double mShooterVelocitySetpoint = 0.0;
+  
+  private double mP = 0.0;
+  private double mI = 0.0;
+  private double mD = 0.0;
+
+  private double mGreenShootingVelocity = 0.0;
+  private double mYellowShootingVelocity = 0.0;
+  private double mBlueShootingVelocity = 0.0;
+  private double mRedShootingVelocity = 0.0;
+
+  public ShooterSubsystem() {
+    mShooterMotor = new MotorController("Shooter Motor", Constants.kShooterMotorPort, Constants.kShooterMotorCurrentLimit);
+    mHoodDoubleSolenoid = new DoubleSolenoid(Constants.kHoodDoubleSolenoidForwardChannel, Constants.kHoodDoubleSolenoidReverseChannel);
+    mShooterPID = mShooterMotor.getSparkMax().getPIDController();
+    mShooterPID.setP(mP);
+    mShooterPID.setI(mI);
+    mShooterPID.setD(mD);
+    SmartDashboard.putNumber("Shooter P Value", mP);
+    SmartDashboard.putNumber("Shooter I Value", mI);
+    SmartDashboard.putNumber("Shooter D Value", mD);
+  }
+
+  // Value.kReverse is when the hood is extended
+  public void setHoodExtended(boolean isExtended) {
+    if (isExtended)
+      mHoodDoubleSolenoid.set(Value.kReverse);
+    else  
+      mHoodDoubleSolenoid.set(Value.kForward);
+  }
+  
+  public boolean getHoodExtendRequired() {
+    return getDistanceFromGoal() > Constants.kHoodExtendRequiredDistance;
+  }
+
+  public void setVelocity(double velocity) {
+    mShooterVelocitySetpoint = velocity;
+    mShooterPID.setReference(velocity, ControlType.kVelocity);
+  }
+
+  public boolean isMotorVelocityWithinRange() {
+    // Compare if the motor velocity is within Constants.kShooterVelocityRange range
+    return (Math.abs(mShooterVelocitySetpoint - mShooterMotor.getEncoder().getVelocity()) < mShooterVelocitySetpoint * Constants.kShooterVelocityPlusMinusPercent);
```

Comment:
For a consistent tolerance, use a percent of the shooter's maximum velocity rather then a percent of the setpoint. 
Currently, this requires a more precise number at lower speeds. 

Additionally, the comment is outdated.

Reply:
Since we are using PID, might be a good idea to take multiple speed measurements to makes sure the speed is stable. 
Could be a good check to make sure the PID is converging and not oscillating.